### PR TITLE
Jackfreem/decoder

### PR DESCRIPTION
**Pull request by jackfreem**
## Description
- Implements the Decoder and DecoderBlock

## Changes Made

- [x]  Builds upon the existing PR with the Encoder and Encoderblock ([link](https://github.com/bradford-gill/frillsformer/pull/10))
- [x] Makes changes to MHA to allow for cross attention

## Checklist

Please ensure the following before submitting the PR:

- [x] PR is Squashed
- [x] I have added/updated relevant documentation (e.g., README.md, inline docstrings).
- [x] I have written tests for my changes.
- [x] All tests pass locally.

## How to Test
`pytest`

### CODE REVIEW

File: tests/blocks/test_decoder_block.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
+import unittest
+import torch
+from src.blocks.decoder_block import DecoderBlock 
+
+class TestDecoderBlock(unittest.TestCase):
+
+    def test_decoder_block_output_shape(self):
```
Comment (bradford-gill) [inline tests/blocks/test_decoder_block.py:7]: hmmm, i asked chatgpt to tell me what to test and the output looks oddly similar lol

Comment (bradford-gill) [reply tests/blocks/test_decoder_block.py:7]: totally fine (and encouraged) to use AI here

File: tests/blocks/test_decoder_block.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        # Mock inputs without masks
+        x = torch.rand(batch_size, seq_length, embed_dim)  # Decoder input
+        encoder_output = torch.rand(batch_size, seq_length, embed_dim)  # Encoder output
+
+        # Forward pass without masks
+        output = decoder_block(x, encoder_output)
+
+        # Check that the output shape matches the input shape
+        self.assertEqual(output.shape, x.shape)
+
```
Comment (bradford-gill) [inline tests/blocks/test_decoder_block.py:46]: can we add a test targeting the mask?
ei:
```python
def test_decoder_block_with_mask():
    batch_size = 2
    tgt_seq_length = 5
    src_seq_length = 10
    embed_dim = 64
    num_heads = 8
    ff_dim = 128

    decoder_block = DecoderBlock(embed_dim, num_heads, ff_dim)
    x = torch.randn(batch_size, tgt_seq_length, embed_dim)
    encoder_output = torch.randn(batch_size, src_seq_length, embed_dim)

    # Masks
    tgt_mask = torch.tril(torch.ones(tgt_seq_length, tgt_seq_length))  # Lower triangular mask
    memory_mask = torch.ones(tgt_seq_length, src_seq_length)

    output = decoder_block(x, encoder_output, tgt_mask, memory_mask)
```

**Review by bradford-gill — APPROVED** [review]
Small changes but looks good to me!

File: tests/blocks/test_decoder_block.py (hunk 1 above)
Comment (jackfreem) [reply tests/blocks/test_decoder_block.py:7]: I bet chatgpt is training on my code
